src/csharp/Grpc.Core/ChannelCredentials.cs

```diff
@@ -171,7 +204,20 @@ internal override bool IsComposable
 
         internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
         {
-            return ChannelCredentialsSafeHandle.CreateSslCredentials(rootCertificates, keyCertificatePair);
+            return ChannelCredentialsSafeHandle.CreateSslCredentials(rootCertificates, keyCertificatePair, this.verifyPeerCallbackInternal);
+        }
+
+        private int VerifyPeerCallbackInternal(IntPtr host, IntPtr pem, IntPtr userData, bool isDestroy)
+        {
+            if (isDestroy)
+            {
+                this.gcHandle.Free();
+                return 0;
+            }
+
+            var context = new VerifyPeerContext(Marshal.PtrToStringAnsi(host), Marshal.PtrToStringAnsi(pem));
+
+            return this.verifyPeerCallback(context) ? 0 : 1;
```

Comment:
you need to catch exceptions here, throwing from within a native callback will crash the application.
See https://github.com/grpc/grpc/blob/8fd43c2e91d6fb4f2cd62b82ec1fe7b2218cd66f/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L71

Reply:
Thanks I fixed it.